### PR TITLE
Native  price estimator using 1Inch spot token API

### PR DESCRIPTION
**Pull request by fleupold**
Creates a native price estimator using the [1Inch spot token API](https://portal.1inch.dev/documentation/spot-price/swagger). It is querying the list of tokens periodically in the background (getting ~1300 per request) and allows fast lookup for this list via a HashMap. The API even allows to query for additional tokens that are not part of this list (e.g. [HarryPotterObamaSonic10Inu](https://etherscan.io/token/0x72e4f9f808c49a2a61de9c5896298920dc4eeea9)) so we can even potentially improve the hit rate by recording requests that weren't in the default list.

The idea for this estimator is to work as a cheap first line of defence (together with the baseline estimator and potentially Quasimodo, cf #1841) to avoid more expensive API requests on native token queries.

It's not yet hooked up, because I'm still wondering if we should use the general `PriceEstimating` trait (which requires additional sanity checks that we are indeed receiving a native price estimate) or the more specific `NativePriceEstimating` trait. 
Currently we do not have a `RacingCompetitionPriceEstimator` component for `NativePriceEstimating` (only `PriceEstimating`) and I'm not sure what the best way is to implement it without repeating this non trivial component. Technically we could use the spot price also to compute generic "estimates" (but they will likely not be accurate) and the interface currently requires shoehorning a zero gas amount and unknown solver which is kind of ugly.

Looking for suggestions!

### Test Plan

Create you own developer account + API Key, then 
```
export ONEINCH_AUTH_TOKEN=...
cargo test --package shared --lib -- price_estimation::native::oneinch::tests::works --exact --nocapture --ignored
```

### CODE REVIEW

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:24]: IIRC the long term plan is to synchronize these kinds of things with the `block_stream` module. That would allow us to only fetch new prices when we detect a new block. This has the potential to return outdated prices when 1Inch did not yet update their prices based on that block but so does a hardcoded interval.
Also hardcoding doesn't play nice with different chains (e.g. `GC` needs `5s`).

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:33]: I think I would drop everything but the `prices` cache from this struct. Because all the requests are done by the background task the estimator only needs to know about the `prices` cache.
Then all the other data could be moved into a separate struct or simply be passed by value into the function that spawns the background task.
This would improve the separation of concerns IMO.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:65]: A bit of context on what failed exactly would be nice when relying just on the logs in Kibana.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:77]: I overlooked this when I simplified the price APIs but it makes more sense for `token` to be passed by value because it's already a `Copy` type. 🤦‍♂️

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:91]: > Currently we do not have a RacingCompetitionPriceEstimator component for NativePriceEstimating (only PriceEstimating) and I'm not sure what the best way is to implement it without repeating this non trivial component.

Since you are already implementing `PriceEstimating` it should be no problem because the `NativePriceEstimator` adapter is currently wrapped around a regular `RacingCompetitionPriceEstimator` using this implementation. So at the moment there is only a single component implementing `NativePriceEstimating`.

Alternatively I think it should be doable to make `RacingCompetitionPriceEstimator` generic over the API (`native` vs `regular`) to make it usable with either estimator type. This could be done by making the inner implementation generic enough to use `impl Fn()` types and have 2 variants of the estimator wrap the core logic (1 for `native`, 1 for `regular`).

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:106]: This should probably use`checked_div` in case the API ever returns `0` for one of the prices.
Also you can already `drop` the `prices` lock before computing `reverse_price`.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (MartinquaXD) [inline crates/shared/src/price_estimation/native/oneinch.rs:215]:
```suggestion
        // These queries are not allowed
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (fleupold) [reply crates/shared/src/price_estimation/native/oneinch.rs:77]: Will refactor separately

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (sunce86) [inline crates/shared/src/price_estimation/native/oneinch.rs:44]: We probably don't need this log, not even under `trace` level.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (sunce86) [inline crates/shared/src/price_estimation/native/oneinch.rs:78]:
```suggestion
        .context("Failed to fetch Native 1inch prices")?
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (sunce86) [inline crates/shared/src/price_estimation/native/oneinch.rs:81]:
```suggestion
        .context("Failed to parse Native 1inch prices")?
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (sunce86) [inline crates/shared/src/price_estimation/native/oneinch.rs:51]: Have you tried to implement background work using `Maintaining` trait?
Seems to fit nicely especially since the `current_block` is only used to indicate when the update needs to be done.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (sunce86) [inline crates/shared/src/price_estimation/native/oneinch.rs:19]: ubernit:

```suggestion
type Token = H160;
type Price = U256;

    prices: Arc<Mutex<HashMap<Token, Price>>>,
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (fleupold) [reply crates/shared/src/price_estimation/native/oneinch.rs:44]: I think it's helpful to know when a backrun job completes (in order to see for instance if we ever update the cache or how outdated the cache was for a specific quote).

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (fleupold) [reply crates/shared/src/price_estimation/native/oneinch.rs:51]: This is definitely possible. The problem with that is that it makes wiring up the price estimator harder. Price estimators are created in `price_estimation/factory` and maintainers are wired up in the top level `main.rs`. We could push a mutable reference to the `maintainers` vec into the factory and make the factory method have side effects, but I thought simply passing down this stream is simpler.
I agree though there is an inconsistency and we should either kill the `Maintaining` trait and make all maintainers manage their own task or replace the other uses of current_block_stream such as the account [balance cache](https://github.com/cowprotocol/services/blob/9bcf411bf87a4e11f515d6969095db61dd832847/crates/shared/src/account_balances/cached.rs#L117).

Comment (nlordell) [reply crates/shared/src/price_estimation/native/oneinch.rs:57]: I think spawning a background task is fine. In fact, we discovered along the way that the `Maintaining` abstraction wasn't great (and might make sense to re-write existing `Maintaining` implementations to also spawn background tasks for updating on new blocks).

I don't remember the exact time in the past where we started moving away from using the `Maintaining` trait, but I think @MartinquaXD and I discussed changing this in the context of "getting rid of the `shared` crate.

**Review by nlordell — APPROVED** [review]
I think the PR is good as is.

Personally, I think it would have been nice to implement this in the `solvers` crate (similar to #1833) to avoid any kind of special `price_estimators` logic for these special cases since, at least AFAIU, it would be nice to also have a way to get external solvers to be able to support similar way of providing token spot prices right? (_If there is no intention to do this at all, then disregard this comment_).

Left some nits, but nothing major blocking the PR (so pre-maturely approved to reduce back-and-forth).

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [reply crates/shared/src/price_estimation/native/oneinch.rs]: Nice catch!

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [inline crates/shared/src/price_estimation/native/oneinch.rs:81]: Just noticed, but this should likely be customizable so that we can funnel it through our egress.

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [inline crates/shared/src/price_estimation/native/oneinch.rs:89]:
```suggestion
    if !response.status().is_success() {
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [inline crates/shared/src/price_estimation/native/oneinch.rs:98]:
```suggestion
        .context("Failed to fetch Native 1Inch prices")?;
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [inline crates/shared/src/price_estimation/native/oneinch.rs:102]:
```suggestion
        .with_context(|_| format!(
            "Failed to parse Native 1inch prices from {response:?}"
        ))?
```

File: crates/shared/src/price_estimation/native/oneinch.rs
Comment (nlordell) [inline crates/shared/src/price_estimation/native/oneinch.rs:99]: Nit: You can define a `Response` type and use `serde_with` with the `DecimalU256` specifier to avoid writing parsing code.

**Review by MartinquaXD — APPROVED** [review]

**Comment by nlordell** [comment]
> Would this require a colocated driver just for native price estimates?

We could reuse the existing deployed driver (The driver:solvers setup is n:m).

> I think I'd prefer keeping some form of builtin estimators

I don't think there is anything inherently wrong with this approach. My argument for moving it into a solver is that:
- It uses existing price estimation logic/code to deal with native price estimates (modulo additional native price estimator configuration so that it uses a different set of estimators)
- Dog-foods our abstraction (if other external teams want to provide token spot prices like we do with the 1Inch API - this ensures our abstractions are correct so they can do the same thing).
- Uniformity in our setup (all prices come from the same place - i.e. solvers), which makes sense from a `#simplicity` perspective.
- If we had missed the potential divide by 0 panic that @MartinquaXD - it would only bring down the price estimator and not the whole orderbook/autopilot process.

Again, this is more of a design decision vs what is better or worse (and making everything a solver **may even prove the be the wrong choice further down the line!**, i.e. this might be bad advice 😅), but it feels more natural with the co-location direction that at least I envisioned.

**Comment by fleupold** [comment]
I see. So it sounds like generally we want all price estimators (even the one we consider core to the protocol) to run in co-located mode. I do think there is a somewhat structural difference between builtin native price estimators and external solvers (in that we don't necessarily want to trust external solvers with native estimates if we don't have to). I also think running a lot of separate micro-services (even with 1 driver, we will at least run `m` solvers) comes with its own complexity.

One issue I see with doing this now is that we are currently not yet running price estimators in colocated mode in prod and thus would block the whole "reduce API request" project on delivering colocated quotes (in fact the whole setup is quite messed up with built-in, legacy and external solvers). So I'd be inclined to merge this as part of the shared crate and port it over to the solvers crate once we got rid of the legacy solvers.

**Pull request merged**
